### PR TITLE
use X11 global mouse selection for middle-click pasting on Linux Desktop

### PR DESCRIPTION
**Pull request by kevinushey**
This PR introduces support, when pasting through middle-click, for the global mouse selection to and from RStudio Desktop on Linux (where supported). This is done by detecting whether a `"paste"` event was issued in response to a middle click, and if so, ensuring that clipboard data from the selection clipboard is used instead.

This should imply that one can select text within an Ace instance, and then middle-click in a separate application to copy + paste that text to the new application (or vice-versa).

### CODE REVIEW

File: src/cpp/desktop/DesktopGwtCallback.cpp [hunk 1]
```diff
@@ -244,6 +255,24 @@ QString GwtCallback::getExistingDirectory(const QString& caption,
...
+    // if this is a change in the selection contents, track it
+    if (mode == QClipboard::Selection)
+    {
+        QClipboard* cb = QApplication::clipboard();
+        QString text = cb->text(mode);
+
+        // when one clicks on an Ace instance, a hidden length-one selection
+        // will sneak in here. explicitly screen those out.
+        if (text == QString::fromUtf8("\x01"))
+            return;
```
Comment (kevinushey) [inline src/cpp/desktop/DesktopGwtCallback.cpp]: Should we restore the previous global mouse selection contents here? (In other words, should we attempt to revert the changes to the clipboard caused by Ace here?)

Comment (jmcphers) [inline src/cpp/desktop/DesktopGwtCallback.cpp]: Do we know *why* Ace does this? I'd leave it unless we know it's a bug or is causing downstream trouble for us (perhaps they're using it as a sentinel of some kind?).

File: src/gwt/src/org/rstudio/studio/client/JavaScriptEventHistory.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+      for (var i = 0, n = events.length; i < n; i++) {
+         body.addEventListener(events[i], handler, true)
+      }
+      
+   }-*/;
+   
+   private void onEvent(NativeEvent event)
+   {
+      EventData eventData = EventData.create(event);
+      queue_.unshift(eventData);
```
Comment (jmcphers) [inline src/gwt/src/org/rstudio/studio/client/JavaScriptEventHistory.java]: This codepath looks like it'll be pretty hot as we're going to run it every time the mouse moves. Does it make sense to use a linked list as a queue rather than an array so we get cheaper insertion/deletion? (GWT has a Queue interface that can be satisfied by a LinkedList)

File: src/cpp/desktop/DesktopGwtCallback.cpp (hunk 1 above)
Comment (kevinushey) [inline src/cpp/desktop/DesktopGwtCallback.cpp]: From what I understand, it's sort of a leaky implementation detail on the Ace side -- there are some comments from https://github.com/ajaxorg/ace/issues/2124, and the Ace maintainer indicates:

> this is a duplicate of #175, it needs rewriting textinput.js to not use PLACEHOLDER which is tricky to do

Behind the scenes, Ace injects the `\x01` into a hidden selection when an Ace instance is activated, and the browser / web engine treats that as a real selection and so overrides the global selection clipboard item. I don't entirely understand the technical reasons why this is done, but I imagine it's done with good reason (perhaps the only way of achieving some behavior in a consistent, cross-platform way?)

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditorMixins.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+   {
+      EventData event = history_.findEvent(new JavaScriptEventHistory.Predicate()
+      {
+         @Override
+         public boolean accept(EventData event)
+         {
+            String type = event.getType();
+            for (String candidate : CANDIDATES)
+               if (type.contentEquals(candidate))
+                  return true;
```
Comment (jmcphers) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditorMixins.java]: Do we need to check which button triggered the event?

File: src/gwt/src/org/rstudio/studio/client/JavaScriptEventHistory.java (hunk 2 above)
Comment (kevinushey) [inline src/gwt/src/org/rstudio/studio/client/JavaScriptEventHistory.java]: Good idea. I tend to default to plain old arrays / vectors just because they are almost always more performant than linked lists (due to cache locality) but it makes sense to be smarter here.

**Comment by JanMarvin** [comment]
Might be a problem on my end (building RStudio with latest Qt): What works for me, is the selection from text in RStudio to `QClipboard::Selection`. I can paste selected lines from RStudio to a different application. What doesn't work is pasting selected lines into RStudio using middle mouse. If  I do this, I always paste what is in `QClipboard::Clipboard`. Anyway thanks a lot for this!

**Comment by kevinushey** [comment]
Thanks @JanMarvin, I'll take a look today and ping you if I can reproduce + figure out what's going on.

**Comment by kevinushey** [comment]
Okay, I've figured out the issue -- GWT actually uses the old-fashioned IE bitfield definition for mouse buttons, and translates the JavaScript native button field to that representation (likely for backwards compatibility).

I missed this and so was comparing buttons with the wrong values. I should be able to have this fixed up soon!
